### PR TITLE
Add char range overload to DotDot operator

### PR DESCRIPTION
**Pull request by IISResetMe**
This PR attempts to allow the DotDot operator (..) to generate
character ranges, in turn allowing shorthand syntax like:

    $AtoZ = -join('A'..'Z')

Since PowerShell has no shorthand syntax for char literals, we
test whether the left-hand-side is of type String, and subsequently
attempts to convert both operands to a char (in accordance with 
PowerShell operator overload behavior in general).

NOTE: 
1) This breaks a number of currently supported range operations, such
as implicit numerical conversion of strings:
'1'..'100'

2) This implementation is fairly naive, making no attempts to assert whether the input characters (or any codepoints in between) fall outside UCS-2 (ie. surrogates). Attempts to expand ranges bound by surrogate pairs are expected to fail and not covered by this implementation (I will leave this an exercise for someone more experienced with unicode hackery than myself)

I was unable to locate existing tests for the integer range function, please advise on where to add those.

### CODE REVIEW

File: src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs
Comment (lzybkr) [inline src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs]: What do you think about returning `char[]` instead?

File: src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs
Comment (lzybkr) [inline src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs]: I think this special case is correctly handled by the code below and it happens rarely, so maybe we should remove it. (And I see you copied the code from the `int` version, so you could remove it there too.)

**Review by lzybkr — CHANGES_REQUESTED** [review]
The code changes look fine - but we need tests.

File: src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs
Comment (IISResetMe) [reply src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs]: Mainly did it this way to keep it similar to the existing code in Compiler.cs. I'm inclined to make a single Range method that takes the operands as Expression's, de-duplicate the loop that generates the actual sequence and then cast the entire thing to char[], int[] or long[] based on the lhs operand before returning. Will update the code later today

File: src/System.Management.Automation/engine/parser/Compiler.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/Compiler.cs]: Will this work:
```
'A'..'Z' | % { ... }
```

File: src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs]: I wonder - is this ASCII Arithmetic work correctly for unicode?

File: src/System.Management.Automation/engine/parser/Compiler.cs
Comment (IISResetMe) [reply src/System.Management.Automation/engine/parser/Compiler.cs]: Yes, this will produce the sequence A (0x41) through Z (0x5A)

File: src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs
Comment (lzybkr) [reply src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs]: It won't work with surrogate pairs because those can't be represented in a `char`, but I don't think that is an interesting scenario.

Comment (iSazonov) [reply src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs]: Since we added `u{} should we detect surrogate pairs and write an error?

Comment (lzybkr) [reply src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs]: I guess we should error on invalid characters (half of the pair). We could also support surrogate pair ranges if we didn't convert the string to char immediately, instead analyzing the string more thoroughly. But I'm not sure I'd block this PR on either of those.

Comment (iSazonov) [reply src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs]: I guess it will work with traditional languages in common scenarios but before I saw issues from China and Japan mans - if the new operator does not matter to them I agree with you.

**Comment by IISResetMe** [comment]
Oops, rebased powershell:master into the branch, looks like all parent commits are in the PR now, how do I clean it up? 😞

**Comment by markekraus** [comment]
@IISResetMe did you only have the 2 commits? if so you could do this (assuming `PowerShell` is the name of your remote branch for the actual PowerShell repo). Make sure you stash or commit any workspace changes before doing this as they will be lost. I'm not a git expert, and maybe there is a better way. So maybe wait until someone more seasoned either "thumbs up" this or tells you just how wrong I am :)


```bash
git fetch
git checkout -b 20171015fix PowerShell/master
git cherry-pick 6e6dfffc2d0150b504585597ad4df997a174867a
git cherry-pick 6d6f20d28157ee49071f018051ac7b35be57bbb0
git checkout -b 20171015backup overload-DotDot-CharRange-1
git checkout overload-DotDot-CharRange-1
git reset --hard 20171015fix
git push --force
```

This will create a new "20171015fix" branch at the current master on this repo, then cherry-picks your 2 commits, then backs up your existing branch to `20171015backup`, and then hard reset your branch to the "20171015fix" branch. This will both "rebase" your branch on to master and add your commits with out the merge mess.  This will likely mess up any comment history here, but I guess it's already some what messed up.

**Review by IISResetMe — COMMENTED** [review]
These were the most basic tests I could think of for the existing Range operator. All examples I could find in the about_* help files pass these.

**Comment by IISResetMe** [comment]
Thanks @markekraus, that seems to have done the trick.

**Comment by TravisEz13** [comment]
I agree with @lzybkr.  We really need tests.

**Comment by TravisEz13** [comment]
Can you make sure `RangeOperator.Tests.ps1` is UTF8 without BOM?  It shows up as binary.

**Comment by IISResetMe** [comment]
@TravisEz13 Done, https://github.com/IISResetMe/PowerShell/blob/34bc59046698d0de41bb5403a6105c7070b38a46/test/powershell/Language/Operators/RangeOperator.Tests.ps1 is UTF-8 (with updated tests) now

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (iSazonov) [inline test/powershell/Language/Operators/RangeOperator.Tests.ps1]: Please add tests for values - `$Range[0] | Should Be 5`

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (iSazonov) [inline test/powershell/Language/Operators/RangeOperator.Tests.ps1]: Please remove the should - next line do the same and Pester can catch a throw there correctly.

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (iSazonov) [inline test/powershell/Language/Operators/RangeOperator.Tests.ps1]: Formatting - please add space after |

Please add type check as in previous test.

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (iSazonov) [inline test/powershell/Language/Operators/RangeOperator.Tests.ps1]: Please remove.

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (iSazonov) [inline test/powershell/Language/Operators/RangeOperator.Tests.ps1]: Previous tests already test "ascending". I think we could test here only descending order for positive and negative numbers. But I believe we should replace some tests with one - "-2 .. 2" and "2 .. -2" - so we check transition through zero too.

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (iSazonov) [inline test/powershell/Language/Operators/RangeOperator.Tests.ps1]: Please remove extra line.

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (iSazonov) [inline test/powershell/Language/Operators/RangeOperator.Tests.ps1]: I think we can remove the test and add type checks in next test.

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (iSazonov) [inline test/powershell/Language/Operators/RangeOperator.Tests.ps1]: Please remove.

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (iSazonov) [inline test/powershell/Language/Operators/RangeOperator.Tests.ps1]: Please use new syntax `u{0x0110}

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (iSazonov) [inline test/powershell/Language/Operators/RangeOperator.Tests.ps1]: Please check values too as above for numbers.

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (iSazonov) [inline test/powershell/Language/Operators/RangeOperator.Tests.ps1]: Please remove extra line.

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (iSazonov) [inline test/powershell/Language/Operators/RangeOperator.Tests.ps1]: Please remove the test - we do the checks above already.

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (iSazonov) [inline test/powershell/Language/Operators/RangeOperator.Tests.ps1]: Is it really [int]? I think we should use values > int.MaxValue.
And please check values too.

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (iSazonov) [inline test/powershell/Language/Operators/RangeOperator.Tests.ps1]: Is it really [int]? I think we should use values > long.MaxValue.
And please check values too.

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (iSazonov) [inline test/powershell/Language/Operators/RangeOperator.Tests.ps1]: Why so much? I think enough `$Range = 1.1d..3.9d`

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (iSazonov) [inline test/powershell/Language/Operators/RangeOperator.Tests.ps1]: Please add new line at EOF.

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (IISResetMe) [reply test/powershell/Language/Operators/RangeOperator.Tests.ps1]: Current implementation doesn't support values > int.MaxValue, intended to add this in a separate branch/PR.

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (IISResetMe) [reply test/powershell/Language/Operators/RangeOperator.Tests.ps1]: The existing numeric range operator only supports int regardless of input type

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (iSazonov) [reply test/powershell/Language/Operators/RangeOperator.Tests.ps1]: Thanks for clarify. Please remove the confuse tests.

**Comment by TravisEz13** [comment]
@lzybkr Can you update your review?

**Comment by TravisEz13** [comment]
@iSazonov Can you indicate if you think this is ready or not (feel free to say if someone else signs off)?

File: src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs]: @IISResetMe Is this addresed?

File: src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs]: @IISResetMe Could you please address the comment in code or add relevant comments in the PR description for future documentation?

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (iSazonov) [reply test/powershell/Language/Operators/RangeOperator.Tests.ps1]: @IISResetMe Please address the comment about the tests for [long] and [bigint].
We should remove its - they don't make sense.

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (iSazonov) [reply test/powershell/Language/Operators/RangeOperator.Tests.ps1]: Closed.

**Comment by iSazonov** [comment]
@TravisEz13 I updated three comment above - after they have been resolved we'll ready to merge.

File: src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs
Comment (IISResetMe) [reply src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs]: @iSazonov I don't feel confident enough in my unicode-fu to attempt to implement the appropriate handling in code for this at the moment, I've updated the PR description to reflect this deficiency instead.

File: src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs
Comment (IISResetMe) [reply src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs]: I've removed the redundant branch in both overloads now. https://github.com/PowerShell/PowerShell/pull/5026/commits/0406cb37f1556b8497325236672819b54da23e53

File: test/powershell/Language/Operators/RangeOperator.Tests.ps1
Comment (IISResetMe) [reply test/powershell/Language/Operators/RangeOperator.Tests.ps1]: You're right, they don't make too much sense at the moment. I've removed them now

**Comment by IISResetMe** [comment]
@iSazonov I believe I've addressed all outstanding comments :-)

File: src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs]: Thanks. 

Closed.

File: src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs]: @IISResetMe Could you please address the comment?

**Comment by iSazonov** [comment]
@IISResetMe Many thanks! We are near to merge. Please address @lzybkr comment above.

File: src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs
Comment (IISResetMe) [reply src/System.Management.Automation/engine/runtime/Operations/NumericOps.cs]: I like the idea, but for now I think we should keep operator output consistent (ie. return `object[]` with individual items of the expected type).

There are are number of builtin operators that return `object[]` in this fashion (`@()`, `-split`, any comparison operator in array/filter mode etc.), changing the behavior should be a concerted effort across all of them IMO

**Comment by IISResetMe** [comment]
@iSazonov Done

**Comment by iSazonov** [comment]
@lzybkr Could you please take another look if you have a bit free time?

**Review by lzybkr — APPROVED** [review]

**Review by iSazonov — APPROVED** [review]

File: src/System.Management.Automation/engine/parser/Compiler.cs
Comment (kborowinski) [reply src/System.Management.Automation/engine/parser/Compiler.cs]: Apparently it *does not work* on the Powershell Core 6.0.0 GA
```powershell
> 'A'..'Z' | % {$_ }
Cannot convert value "A" to type "System.Int32". Error: "Input string was not in a correct format."
At line:1 char:1
+ 'A'..'Z' | % {$_ }
+ ~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidArgument: (:) [], RuntimeException
+ FullyQualifiedErrorId : InvalidCastFromStringToInteger
```

Comment (markekraus) [reply src/System.Management.Automation/engine/parser/Compiler.cs]: @kborowinski there is an open issue for that #5519

Comment (kborowinski) [reply src/System.Management.Automation/engine/parser/Compiler.cs]: @markekraus: sorry, I've missed that issue

**Pull request merged**
